### PR TITLE
[TS-SELENIUM] Cover the "Use Java IDE features and the inner loop" step from "Happy path" scenario

### PR DESCRIPTION
**Pull request by Ohrimenko1988**
### What does this PR do?
Cover the "Use Java IDE features and the inner loop" step from "Happy path" scenario

Screenshot from 2019-07-01 15-21-16

### What issues does this PR fix or reference?
Issue: https://github.com/eclipse/che/issues/13519

### CODE REVIEW

**Review by musienko-maxim — APPROVED** [review]

**Comment by Ohrimenko1988** [comment]
ci-build

**Review by l0rd — APPROVED** [review]

**Comment by Ohrimenko1988** [comment]
ci-build

**Comment by Ohrimenko1988** [comment]
ci-build

File: e2e/tests/e2e_happy_path/HappyPath.spec.ts
Comment (Katka92) [inline e2e/tests/e2e_happy_path/HappyPath.spec.ts]: Is the skip there intended?

File: e2e/tests/e2e_happy_path/HappyPath.spec.ts
Comment (Katka92) [inline e2e/tests/e2e_happy_path/HappyPath.spec.ts]: Why are those comments left there?

File: e2e/pageobjects/ide/Editor.ts
Comment (rhopp) [inline e2e/pageobjects/ide/Editor.ts:269]: Do we know what errors can happen here? Shouldn't we at least log them?

File: e2e/pageobjects/ide/Editor.ts
Comment (rhopp) [inline e2e/pageobjects/ide/Editor.ts:283]: Maybe using regex would be nicer here? I'm not sure how regexes work in typescript... I've just tried som online javascript like https://playcode.io/online-javascript-editor and this seems to work ok:
```
var str = 'position:absolute;top:627px;width:100%;height:19px;';
console.log(/(?<=top:)(\d*)/.exec(str)[0]); \\prints 627
```

File: e2e/utils/DriverHelper.ts
Comment (Katka92) [inline e2e/utils/DriverHelper.ts:242]: Can you please elaborate why await with timeout is not suitable and you have to use for loop here?

File: e2e/pageobjects/ide/Ide.ts
Comment (rhopp) [inline e2e/pageobjects/ide/Ide.ts:139]: I don't really get this *for* cycle... It seems to me you just want to increase timeout for the `wait`. So why not just increase the timeout?

File: e2e/pageobjects/ide/Terminal.ts
Comment (rhopp) [inline e2e/pageobjects/ide/Terminal.ts]: Again... couldn't we rewrite these sleeps/waits to be "active" wait? What does it mean for the element to be interactable? Couldn't we write this down as a code which waits for this?

File: e2e/pageobjects/ide/Terminal.ts
Comment (rhopp) [inline e2e/pageobjects/ide/Terminal.ts]: Same thing...

File: e2e/pageobjects/ide/TopMenu.ts
Comment (rhopp) [inline e2e/pageobjects/ide/TopMenu.ts:40]: I guess this is done because element with locator `buttonLocator` is not clickable at the time... WDYT about calling `closeAllNotifications`  only when this happens? I think it would make sense to leave this as is if the notifications will be there in 100% time. Is this the case?

File: e2e/pageobjects/ide/WarningDialog.ts
Comment (rhopp) [inline e2e/pageobjects/ide/WarningDialog.ts:28]: nitpick: `dialogIsDisplayed` could be a nicer name for the var imho.

File: e2e/tests/e2e_happy_path/HappyPath.spec.ts
Comment (rhopp) [inline e2e/tests/e2e_happy_path/HappyPath.spec.ts]: I know this is out of scope of this PR, but for some next time - `Abcence` doesn't exist - it's `Absence`

File: e2e/tests/e2e_happy_path/HappyPath.spec.ts
Comment (rhopp) [inline e2e/tests/e2e_happy_path/HappyPath.spec.ts:267]: I guess this method is supposed to live elsewhere, right?

File: e2e/tests/e2e_happy_path/HappyPath.spec.ts
Comment (rhopp) [inline e2e/tests/e2e_happy_path/HappyPath.spec.ts]: Some forgotten commented code. Pls remove this from the final PR.

File: e2e/utils/DriverHelper.ts
Comment (rhopp) [inline e2e/utils/DriverHelper.ts:242]: I know you have already lot of methods like this, where you implemented your polling mechanism, but is this really needed? Couldn't we just use https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/lib/webdriver.js#L448? (and as allways... I'm pretty new to this tech, so there may be reasons for this to be not possible... If yes, please tell me the reasons - I want to get wiser :-D)

File: e2e/pageobjects/ide/Terminal.ts
Comment (rhopp) [inline e2e/pageobjects/ide/Terminal.ts:65]: I guess this is some hack to workaround the fact, that terminal is implemented as canvas? Correct?

**Review by rhopp — COMMENTED** [review]
I've put some comments in here... Most of them are not blocking the PR, but I would like to have some reaction to them before merging the PR.

Also what I'm missing, when reading and - more importantly - running the tests is proper logging. I would like to know what is happening... The most obvious examples are the waiting functions... I would like to see in the logs what is currently being executed, how long are we waiting for different stuff etc. But this is something we would need to tackle in some other PR.

File: e2e/pageobjects/ide/Editor.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/Editor.ts]: The ```selectTab``` method has been already implemented. Just put it here.

File: e2e/utils/DriverHelper.ts
Comment (Ohrimenko1988) [reply e2e/utils/DriverHelper.ts:242]: Because timeout assumes that wait will be failed when the timeout end. And try may be not repeated if waiting for the element is too long. In this case, I want to ensure that try will be repeated.

File: deploy/kubernetes/helm/che/templates/configmap.yaml
Comment (Ohrimenko1988) [reply deploy/kubernetes/helm/che/templates/configmap.yaml:117]: Without this variables each launch fails with "heap space" error

File: e2e/pageobjects/ide/DebugView.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/DebugView.ts]: In this case, it is not critical because it works properly even without this sleep. This is my overcautiousness for CI launches

File: e2e/pageobjects/ide/Editor.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/Editor.ts]: Yup, I used this variable for debug logging, logging was removed and the variable is left

File: e2e/pageobjects/ide/Editor.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/Editor.ts:269]: In this case, we should ignore all exceptions

File: e2e/tests/e2e_happy_path/HappyPath.spec.ts
Comment (Ohrimenko1988) [reply e2e/tests/e2e_happy_path/HappyPath.spec.ts:267]: In my opinion no, because this method is too specific for the curent test

File: e2e/utils/DriverHelper.ts
Comment (Ohrimenko1988) [reply e2e/utils/DriverHelper.ts:242]: Similar question and ansver https://github.com/eclipse/che/pull/13615#discussion_r299926556 .
And a small addition. Most of the mentioned methods really maybe reworked by waits. I have already done something similar  https://github.com/eclipse/che/blob/master/e2e/utils/DriverHelper.ts#L150-L170
It may be done in the separate PR

File: e2e/pageobjects/ide/Terminal.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/Terminal.ts:65]: Not exactly, the terminal similar to the editor has specific web element 6х6 pixels which are responsible for interaction with editor or terminal. In the terminal case, this web element is not visible.

File: e2e/utils/DriverHelper.ts
Comment (rhopp) [reply e2e/utils/DriverHelper.ts:242]: So the possible solution could be reduce the timeout for waiting for the elements inside the "waiting function" and increasing timeout for the wait, right? But yeah... As you mentioned under my comment here.. this could be done in other PR.

File: deploy/kubernetes/helm/che/templates/configmap.yaml
Comment (rhopp) [reply deploy/kubernetes/helm/che/templates/configmap.yaml:117]: I may have not been completely clear here... I was talking just about `CHE_INFRA_KUBERNETES_WORKSPACE__START__TIMEOUT__MIN`. This one is not affecting command execution (build inside workspace) in any way.

File: e2e/pageobjects/ide/DebugView.ts
Comment (rhopp) [reply e2e/pageobjects/ide/DebugView.ts]: I strongly believe any sleep like this should be either removed, or replaced by conditional wait.
These three waits in this method are perfect example of this... You even put into comments what should we wait for.
This is the strongest reason preventing me from giving approval to this PR. If it works OK without them (even on CI), we could remove them. If it doesn't work right without them, we should replace them with conditional waits.

File: e2e/pageobjects/ide/Terminal.ts
Comment (rhopp) [reply e2e/pageobjects/ide/Terminal.ts:20]: Editor & terminal are two different "parts of IDE"... they may have these methods in common, but I don't think it would be wise to abstract this (make a common ancestor for these components), as these components are not very related....

File: e2e/pageobjects/ide/Terminal.ts
Comment (rhopp) [reply e2e/pageobjects/ide/Terminal.ts:65]: Thanks for explanation!

File: e2e/pageobjects/ide/Terminal.ts
Comment (rhopp) [reply e2e/pageobjects/ide/Terminal.ts]: I guess this is still something I strongly feel about... Until I'm proven we **really need** this and it couldn't be done any other way I would request this to change.

**Review by rhopp — CHANGES_REQUESTED** [review]
I've put some more comments here... I'm also requesting changes, as I feel strongly about "hard sleeps" ( like `driverHelper.wait(2000)`).

File: e2e/files/happy-path/happy-path-workspace.yaml
Comment (Ohrimenko1988) [reply e2e/files/happy-path/happy-path-workspace.yaml]: Already changed locally to the branch URL

File: e2e/pageobjects/ide/Editor.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/Editor.ts:269]: Because actions inside of try/catch are very sensitive, and I tried to process most common exceptions and final was that the "WebDriverError" may be happening and it does not make sense to process other exceptions because it is a parent of all driver exceptions. Anyway, if something went wrong exception with full explanation will be thrown in the further code.

File: e2e/pageobjects/ide/Editor.ts
Comment (Ohrimenko1988) [reply e2e/pageobjects/ide/Editor.ts:109]: It is a bit hard to implement. If we have the easiest way why not use it? It is also reliable.

Comment (rhopp) [reply e2e/pageobjects/ide/Editor.ts:109]: Makes sense. Resolving this comment

**Review by rhopp — APPROVED** [review]
There are still some questions/nitpicks unanswered, but all of them are of minor character.

**Pull request merged**
